### DISCUSSION TITLE
How to include multiple .appxsym files in a .msixupload file?

### DISCUSSION THREAD

**Discussion by myfix16** [microsoft/WindowsAppSDK#3471] [Q&A] [answered]
Hi. I have a packaged WASDK application and I want to publish it on Microsoft Store. According to [this guide](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/single-project-msix?tabs=csharp), 

> Single-project MSIX doesn't currently support producing MSIX bundles (see [Bundling MSIX packages](https://learn.microsoft.com/en-us/windows/msix/package/bundling-overview)). It produces only a single MSIX.

Therefore, I manually create the .msixbundle using `MakeAppx.exe`. Then, as I've read [here](https://github.com/MicrosoftDocs/msix-docs/blob/master/msix-src/package/packaging-uwp-apps.md#generate-an-app-package), to create a `.msixupload` file I need a `.appxsym` file. The question is that I have three `.appxsym` files for each package (x86, x64, and Arm64), and I wonder:
1. Should I include any one of the `.appxsym` files or include all of them?
2. If I have to do the latter, where should I put these files, and how should I name them?

Disclaimer: A majority part of this is referenced from [this question](https://techcommunity.microsoft.com/t5/msix/how-to-create-msixupload-file-manually/m-p/2964813)

**Comment by DarranRowe** [comment, marked as answer]
Since this issue only occurs with the single project MSIX, there is one option available to see what Visual Studio itself does.
Visual Studio has project templates that have separate application packaging projects (Blank App, Packaged with Windows Application Packaging Project (WinUI 3 in Desktop) for both C++ and C#). These projects do support creating MSIX bundles. You can create one of these projects that has the default application, make a sideloading package and bundle it, then look at what it does for the symbol files. This should work with the single project output too.

**Reply by myfix16** [reply]
Thanks. I tried this method and found that you should put all .appxsym files in the .msicupload file.

**Reply by steam3d** [reply]
When I try create a single-project bundle for Microsoft Store I got a lot files. I use [this](https://github.com/MicrosoftDocs/msix-docs/blob/main/msix-src/package/packaging-uwp-apps.md#create-your-app-package-upload-file-manually) to create a bundle manually. I zipped x86,x64,ARMx64 architectures but when I upload it to Microsoft Store, the package shows as only ARM64. 

So I could not create a bundle and could not find information what to do.

**Reply by DarranRowe** [reply]
Well, since appx/msix are basically zip files, the one thing you can do for sanity testing is go to where your .msixbundle file is, copy it and rename the copy to have a .zip extension, then finally check the contents.
As an example, I have a project named BasicWindowTest, in this case, it is a non WinUI desktop application packaged up with a separate packaging project, but it helps illustrate how things work.
When it finishes creating the bundle, there is a file named BasicWindowTestPackage_1.0.0.0_x86_x64_arm64.msixbundle. This file name itself isn't really relevant besides the extension. Anyway, I copy this file and make sure the name of the copied file is BasicWindowTestPackage_1.0.0.0_x86_x64_arm64.zip. The file contains the following:

![Screenshot 2023-03-10 212021](https://user-images.githubusercontent.com/52577874/224430620-bc7bf51a-0748-450d-8ceb-86be0ae9d3ae.png)

As you can see, the bundle contains the three .msix files. So the sanity check you should do is make sure that your generated .msixbundle contains all of your .msix files. The way Visual Studio generates the bundle is

1) Copy all of the generated .msix files to a directory.

![Screenshot 2023-03-10 212813](https://user-images.githubusercontent.com/52577874/224431896-45e6ecd6-5c34-4525-bd53-3ab5a7ebda88.png)

It using the ARM64 directory here isn't important. I think it just does that due to it being the last platform built.

2) It runs makeappx to generate the bundle. For this project, Visual Studio used:
`C:\Users\Darran\.nuget\packages\microsoft.windows.sdk.buildtools\10.0.22621.755\bin\10.0.22621.0\x64\MakeAppx.exe bundle /d obj\ARM64\Release\BasicWindowTestPackage_1.0.0.0_Bundle\ /bv 1.0.0.0 /o /p C:\Users\Darran\source\repos\BasicWindowTest\BasicWindowTestPackage\AppPackages\BasicWindowTestPackage_1.0.0.0_Test\BasicWindowTestPackage_1.0.0.0_x86_x64_arm64.msixbundle`
on my system. As you can see, the input is the directory where it placed all of the .msix files and the output is the .msixbundle file.

3) It then signs the bundle with the same certificate that it used for the individual .msix files.

**Reply by steam3d** [reply]
I figure out. Yep I could create the *.msixbundle without symbols. But why bundle does not support symbols. 

Did you simple zipped 3 files .appxsym and BasicWindowTestPackage_1.0.0.0_x86_x64_arm64.msixbundle to *.msixupload?

**Reply by steam3d** [reply]
Yes. It  works. Microsoft Store detects package correct. I am not sure how to check symbols, but *.msixupload contains them
![image](https://user-images.githubusercontent.com/30986874/224439630-c66b5304-017e-4529-9278-e577133837ce.png)

I will trying to submit real project to Microsoft Store and return with feedback.

**Reply by myfix16** [reply]
> I figure out. Yep I could create the *.msixbundle without symbols. But why bundle does not support symbols.
> 
> Did you simple zipped 3 files .appxsym and BasicWindowTestPackage_1.0.0.0_x86_x64_arm64.msixbundle to *.msixupload?

@steam3d  Yes, I tried it with my real project, and it worked perfectly.

**Reply by myfix16** [reply]
I zipped 3 `.appxsym` files and the bundle in the `.msixupload` file, and published my app on the Microsoft store. However, in the health page of partner center, I found only a part of the failure is recognized, and the remaining is recognized as "Unknown". Do anyone encounter the same issue?
![image](https://user-images.githubusercontent.com/61649477/224622048-f12fd20f-351d-4977-bd08-3099ff9f735f.png)

**Reply by steam3d** [reply]
Yes. I have the same result even if Visual Studio creates bundle itself. I am not sure that Unknown is real crushes. For 2 years in Microsoft Store, I could not figure out what is Unknown.

**Reply by eskilgh** [reply]
I have the same problem. We have an app created with .NET MAUI, that is built to a .msix file and a separate .msixsym file containing debug symbols. I changed the extension from .msixsym to .appxsym (as suggested [here ](https://github.com/microsoft/WindowsAppSDK/discussions/1952)). Then, following [these instructions ](https://learn.microsoft.com/en-us/windows/msix/package/packaging-uwp-apps#create-your-app-package-upload-file-manually) compressed the .msix and .appxsym files in and archive and changed the extension to .msixupload.

Uploading the .msixupload to store works fine: The package is certified and published, and installing it from the store works just as before. However, I can't see that any of the symbols have been unpacked after installing? I checked the %ProgramFiles%/WindowsApps/<my app name> folder, but found none of the pdb files from the .msixsym/.appxsym file. Are the symbols unpacked at another destination?

I have to say, this is incredibly frustrating. Having reliable crash reports is a crucial part of app development.

**Reply by steam3d** [reply]
oh shıt here we go again.  Now I faced with localization issue. I have the project that use external app package and after build it create a single `.msixbundle` and it has each localization language as separate `.msix` file. It looks like this:
![image](https://github.com/microsoft/WindowsAppSDK/assets/30986874/1f0d691e-680a-4a06-aa01-93c3608d0327)

The Microsoft Store understands that package containt multiple language and automaticly add them to store lintising.  So I have no idea how to build self-packaged Winui3 app like this.

**Reply by DarranRowe** [reply]
This is something that the separate .wapproj does, so you can reference that.
In general, it goes:
1) Generate your full .pri file using makepri.
2) Split your .pri file using makepri based upon criteria that you give it.
3) Create the main .appx/.msix package using makeappx. You use the output of the split resources above for the main package.
4) Create all of the resource packages using makeappx. Again, you use the output of the split resources above.
5) Sign all of the packages with your certificate, this includes all of the resource packages.

This can be a bit of a pain to put together, but it is possible to do it purely on the command line.
